### ISSUE TITLE
Can't open app bundle on macOS (app is "damaged" and can't be opened)

### ISSUE THREAD

**Issue by codecat** [deskflow/deskflow#7606]
### Project

Deskflow

### Deskflow version number

1.17.0+r1

### Build type

Deskflow package (downloaded from us)

### Operating systems (OS)

- [ ] Windows
- [X] macOS
- [ ] Linux
- [ ] BSD-derived
- [ ] Other

### OS versions/distros

- MacOS 15.0

### Deskflow configuration

- Windows 11
- MacOS

### What steps will reproduce the problem?

1. Install `Deskflow.app`
2. Run app

<img width="274" alt="image" src="https://github.com/user-attachments/assets/23e1d6ee-e922-47cf-a6ea-0a515f366057">

### Additional information

I guess the build is currently not signed or notarized. Can't find much info so figured I'd make an issue to track it.

**Comment by sithlord48** [comment]
Go to system settings and under security you will have a run anyway option for it show up for some time after this dialog.. Apples really into making sure you pay them for a valid cert..

**Comment by codecat** [comment]
Thanks. I don't see the "Run anyway" option currently, but maybe I have to wait a bit.

Is the plan to get it signed & notarized in the future?

**Comment by nbolton** [comment]
Take a look at: [Running on macOS](https://github.com/deskflow/deskflow/wiki/Running-on-macOS)

I'm working with Homebrew on getting a Deskflow package so that'll be the official way to install it, since that takes care of all the signing problems nicely.

**Comment by codecat** [comment]
Thank you! Unfortunately, the `spctl` command doesn't appear to work anymore as of MacOS 15, as it throws the following error:
```
This operation is no longer supported. Please see the man page for more information.
```

The man page displays this:

<img width="993" alt="image" src="https://github.com/user-attachments/assets/0a6898d0-a9d5-4d67-bb69-c2bc46f94bcd">

**Comment by nbolton** [comment]
> Unfortunately, the `spctl` command doesn't appear to work anymore as of MacOS 15

Gah, Apple. Moving target.

Did this command work?
```
xattr -dr com.apple.quarantine /Applications/Deskflow.app
```

I'm like 90% sure that we only need the `xattr` but I haven't done thorough testing on that.

**Comment by nbolton** [comment]
> And I don't see anything in the Security settings.

Do you see the 2nd `deskflow` item? Could you share a screenshot of what you do see?

![Uploading image.png…]()

**Comment by codecat** [comment]
> Did this command work?
> 
> ```
> xattr -dr com.apple.quarantine /Applications/Deskflow.app
> ```

Ah, it did! After starting it again, it gave me the option to enable it in the settings and get it to run. 🎉

**Comment by Dentrax** [comment]
![Screenshot 2024-10-02 at 23 18 20](https://github.com/user-attachments/assets/1f1b5ab3-fce8-45ea-a8b5-cae258a57918)
![Screenshot 2024-10-02 at 23 18 53](https://github.com/user-attachments/assets/df64ec54-77d4-41aa-9845-0f0b364b68d0)

P.S: Restarting didn't work.

**Comment by codecat** [comment]
@Dentrax I opened `Deskflow.app` using "Show app contents", navigated to where the `deskflow` binary is, and dragged it into the accessibility settings list. It works now! 🚀

**Comment by codecat** [comment]
To clarify the steps I did to make it work on MacOS 15, in case it helps someone (or for the wiki page):

I ran: `xattr -dr com.apple.quarantine /Applications/Deskflow.app`

Then I opened the app contents: `open /Applications/Deskflow.app/Contents/MacOS/`

And dragged the `deskflow` binary to the window:

<img width="505" alt="image" src="https://github.com/user-attachments/assets/2a61325a-89b7-40fc-80c5-e84fb82d6c1b">

**Comment by Dentrax** [comment]
It worked! Thank you @codecat! ❤️ 

I think this should be also documented on the wiki.

**Comment by nbolton** [comment]
> @Dentrax I opened `Deskflow.app` using "Show app contents", navigated to where the `deskflow` binary is, and dragged it into the accessibility settings list. It works now! 🚀

Very nice, thanks for sharing!

> I think this should be also documented on the wiki.

[Done!](https://github.com/deskflow/deskflow/wiki/Running-on-macOS)

**Comment by Dentrax** [comment]
Thanks, small nit:

```diff
- Drag the deskflow binary to the accessibility window.
+ Drag all of the deskflow-* binaries to the accessibility window.
```

**Comment by sithlord48** [comment]
This is 100% an Apple issue. Perhaps if Apples customers made reports to Apple about how they keep making it harder to run unsigned apps. What was wrong with the dialog asking if you wanted to run it anyway ??? 

Tldr: The "bug" is we don't pay Apple $100/yr to have a cert registered with them.

**Comment by nbolton** [comment]
> This is 100% an Apple issue.

It is what it is. I'm happy to take this on as my responsibility to fix. I'm trying to decide if it's necessary to fork out for signing (what they want you to do) or if we can circumvent the status quo by having self-signed be our official solution.

**Comment by sithlord48** [comment]
Its the lack of signing that is a problem. it is signing with a key that is on the appstore keyring (what you pay for). Apple doesn't care that that it is signed only that you pay them to be on the keyring. If your key is not on there users need to import our key a different hurdle for them to overcome. (but not impossible)

**Comment by codecat** [comment]
Well, whatever your opinion is on Apple's signing requirements, at the end of the day you need to have a reliable and easy way to distribute the app to your users. It sounds like the Homebrew way may be the best way if you don't want to pay Apple and also don't really care about distributing it elsewhere.

**Comment by sean5446** [comment]
Thanks for starting this thread. I also saw this error as well installing on Mac. I didn't realize the error, "Deskflow.app is damaged and can't be opened" was Mac-speak for a security permission issue. I opted to try the x86 build despite running on an M2 ARM Mac and got a 'normal' looking error along the lines of "This app is untrusted" instead of "This app is broken".

**Comment by sithlord48** [comment]
Since Mac os15 All unsigned apps are considered `damaged` you can still ignore that and tell the system to open anyway you just don't have a prompt and the message does not make it clear its a signing issue. Apple doesn't address the dialog but they did update the  [Help Topic](https://support.apple.com/guide/mac-help/open-a-mac-app-from-an-unknown-developer-mh40616/mac) on opening apps from unknown developers for 10.15.

**Comment by nbolton** [comment]
> The option -r is not recognized.

Thanks for reporting. Updated wiki page: [Running on macOS](https://github.com/deskflow/deskflow/wiki/Running-on-macOS)

**Comment by sithlord48** [comment]
> In my case, It doesn't work even after I add the firewall settings, so I have to disable the firewall for now.
> 
> macOS 15.0.1

I can confirm it at least is working on my system with 15.0.1

Some oddities i've noticed about that setting on mac os
 - If you don't remove it before you run a new version it get tripped up allowing the new version of the app
 - you need to allow it once to run the gui another time to run the client and another time to run the server resulting in upto 3 entries in the accessibility

**Comment by nbolton** [comment]
> have to disable the firewall

I think this is a separate issue. Would you mind opening a new bug for that?

**Comment by EliW** [comment]
> For now, take a look at: [Running on macOS](https://github.com/deskflow/deskflow/wiki/Running-on-macOS)
> 
> I'm [working with Homebrew](https://github.com/deskflow/deskflow/discussions/7603) on getting a Deskflow package so that'll be the official way to install it, since that takes care of all the signing problems nicely.

Hey @nbolton Any update on getting this working via homebrew?   Looking to use DeskFlow again (well, was Synergy Community decades ago when I used it last) ... and it would be much more palatable to corporate entities if this just was a nicely-signed-worked directly-installed-via-brew solution.   Versus needing to go hack into the quarantine level stuff.

**Comment by rexzhang** [comment]
> > have to disable the firewall
> 
> Would you mind opening a new bug for that? It's a separate issue.
> 
> The macOS 15 Sequoia firewall problems are separate to this issue and related to:
> 
>     * https://symless.com/synergy/help/macos-15-sequoia-firewall-breaks-network-connection
> 
>     * https://forums.symless.com/topic/7704-loss-of-connection-and-focus-with-macos-sequoia/
> 
>     * https://mjtsai.com/blog/2024/09/18/macos-firewall-regressions-in-sequoia/
> 
>     * https://techcrunch.com/2024/09/19/apples-new-macos-sequoia-update-is-breaking-some-cybersecurity-tools/
> 
>     * https://www.theregister.com/2024/09/23/security_in_brief/
> 
>     * https://discussions.apple.com/thread/255765230?sortBy=rank
> 
>     * https://uk.pcmag.com/security/154463/macos-sequoia-causing-vpn-web-browsing-antivirus-software-problems
> 
>     * https://support.apple.com/en-us/120283

Sorry about this, I'm not sure it's a Deskflow bug or a mac OS problem. 
I've tried different solutions, including and not limited to: manually adding firewall policies, local network permissions, and so on.

I'm back to Synergy3 for now. If there are still problems after mac OS update, I'll come back and open a new issue.

**Comment by Dentrax** [comment]
I can confirm this. I didn't see that Firewall issue on MacOS 15 and I just upgraded... Eventually Deskflow completely broken.

I was following the [Synergy's doc](https://symless.com/synergy/help/macos-15-sequoia-firewall-breaks-network-connection) but there is no such `deskflow` apps under `Local Network` privacy settings.

Deskflow server logs are:
```
[2024-10-23T14:50:41] INFO: starting core server process (desktop mode)
[2024-10-23T14:50:41] INFO: core config file: /Users/furkan/Library/Preferences/Deskflow/deskflow-server.conf
[2024-10-23T14:50:41] INFO: running command: /Applications/Deskflow.app/Contents/MacOS/deskflow-server -f --no-tray --debug INFO --name Furkans-MacBook-Pro.local --address :24800 -c /Users/furkan/Library/Preferences/Deskflow/deskflow-server.conf
[2024-10-23T14:50:41] NOTE: clipboard sharing is disabled
[2024-10-23T14:50:41] NOTE: started server, waiting for clients
2024-10-23 14:50:41.279 deskflow-server[13188:81738] starting cocoa loop
```

I don't think this is a Deskflow bug. Is there anything else that we can do besides waiting a fix version of macOS? Thank you Apple...

**Comment by nbolton** [comment]
> I can confirm this. I didn't see that Firewall issue on MacOS 15 and I just upgraded... Eventually Deskflow completely broken.

Please add your Sequoia firewall comments to: #7794

This issue is (#7606) related to package signing. I will update the title of this issue to be more accurate.

**Comment by aoberoi** [comment]
(this might not be the right issue for this)

I just performed the Homebrew install and it worked well, except I have  one suggestion to improve the instructions in the README. After installing, I opened Deskflow.app and got a dialog regarding permission for accessibility control. I used the button in that dialog to open the correct pane in System Settings, but the Deskflow app was there and the toggle was already turned on. So I manually quit Deskflow.app and restarted. I once again got the same prompt. At this point I was a bit lost.

Then I read here that someone dragged the 3 binaries from inside the Deskflow.app bundle into the System Settings window and enabled each of them for accessibility controls. Then I relaunched the Deskflow app and it worked.

I don't know if all 3 binaries are required, but my suggestion is to add details about dragging the binaries in and toggling the switch to grant the permission.

**Comment by Lithobius** [comment]
> Thanks for starting this thread. I also saw this error when installing on Mac. I didn't realize the error, "Deskflow.app is damaged and can't be opened" was Mac-speak for a security permission issue. I opted to try the x86 build despite running on an M2 ARM Mac and got a 'normal' looking error along the lines of "This app is untrusted" instead of "This app is broken" and was able to proceed through the regular hoops Mac makes you jump through to install software.

This worked for me as well today to connect an M1 and intel mac both on 13.7.2

**Comment by sithlord48** [comment]
> (this might not be the right issue for this)
> 
> I opened Deskflow.app and got a dialog regarding permission for accessibility control. I used the button in that dialog to open the correct pane in System Settings, but the Deskflow app was there and the toggle was already turned on. So I manually quit Deskflow.app and restarted. I once again got the same prompt. At this point I was a bit lost.

That seams to be a bug in macOS since have this issue with not only accessibility but other permissions and with a few apps on upgrade. If its there remove the app using the - button in the panel and restart the app. If you are not able to remove it with that button then uninstall the app and restart the machine that should reset the permission. I had this happen with a few apps from various install sources.

**Comment by joyrider3774** [comment]
if you ad hock code sign the app bundle, the damaged message should be replaced with a can't validated for authenticity message. The thing is with the damaged app message users can not let it run through security and privacy but if you ad-hoc code code sign your app bundle while not having an apple developper id it seems to function fine. I've tried it on my own mac builds done in github runners as well as on other. For more detailed info about this check this is issue i recently created https://github.com/Ravbug/sdl3-sample/issues/19

**Comment by sithlord48** [comment]
We do sign locally but deep signing is not recommend anymore. This has been resolved since then
